services/worldstatestore/client.go

```diff
@@ -0,0 +1,156 @@
+	ctx context.Context,
+	conn rpc.ClientConn,
+	remoteName string,
+	name resource.Name,
+	logger logging.Logger,
+) (Service, error) {
+	grpcClient := pb.NewWorldStateStoreServiceClient(conn)
+	c := &client{
+		Named:  name.PrependRemote(remoteName).AsNamed(),
+		name:   name.ShortName(),
+		client: grpcClient,
+		logger: logger,
+	}
+	return c, nil
+}
+
+// ListUUIDs lists all UUIDs in the world state store.
+func (c *client) ListUUIDs(ctx context.Context, extra map[string]interface{}) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "worldstatestore::client::ListUUIDs")
+	defer span.End()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &pb.ListUUIDsRequest{Name: c.name, Extra: ext}
+	resp, err := c.client.ListUUIDs(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	uuids := resp.GetUuids()
+	if uuids == nil {
+		return nil, ErrNilResponse
+	}
+
+	return uuids, nil
+}
+
+// GetTransform gets the transform for a given UUID.
+func (c *client) GetTransform(ctx context.Context, uuid []byte, extra map[string]interface{}) (*commonPB.Transform, error) {
+	ctx, span := trace.StartSpan(ctx, "worldstatestore::client::GetTransform")
+	defer span.End()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &pb.GetTransformRequest{Name: c.name, Uuid: uuid, Extra: ext}
+	resp, err := c.client.GetTransform(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	obj := resp.GetTransform()
+	if obj == nil {
+		return nil, ErrNilResponse
+	}
+
+	return obj, nil
+}
+
+// StreamTransformChanges streams transform changes.
+func (c *client) StreamTransformChanges(ctx context.Context, extra map[string]interface{}) (<-chan TransformChange, error) {
+	ctx, span := trace.StartSpan(ctx, "worldstatestore::client::StreamTransformChanges")
+	defer span.End()
+
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &pb.StreamTransformChangesRequest{Name: c.name, Extra: ext}
+	stream, err := c.client.StreamTransformChanges(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	// Check the initial response immediately to catch early errors.
+	_, err = stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+
+	changesChan := make(chan TransformChange, 1024)
+
+	go func() {
+		defer close(changesChan)
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+				}
+				if ctx.Err() != nil || errors.Is(err, context.Canceled) {
+					c.logger.Debug(err)
+					return
+				}
+				c.logger.Errorw("failed to receive from stream", "error", err)
+				return
+			}
+
+			change := TransformChange{
+				ChangeType: resp.ChangeType,
+				Transform:  resp.Transform,
+			}
+
+			if resp.UpdatedFields != nil {
+				change.UpdatedFields = resp.UpdatedFields.Paths
+			}
+
+			select {
+			case changesChan <- change:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return changesChan, nil
+}
```

Comment:
@benjirewis, you don't need to review everything, but mind taking a look at how we are managing this stream to make sure there are no glaring issues? I tried a few things, and it is shutting down cleanly and working e2e.

Reply:
I took a look here and in the server code and compared it to some server-streaming code I've seen; I see no glaring issues. Returning a channel seems pretty ergonomic to me, but I suppose you could also return a struct that implements `Next() (TransformChange, error)`.

Also cc @viamrobotics/netcode since a whole new service type is being added here with a non-trivial amount of code. We'll have to change some small pieces of this (`ShortName()`) @jmatth.

Reply:
OK, returning a stream struct looks a little cleaner than returning the channel directly.